synapse/handlers/federation_event.py

```diff
@@ -661,10 +757,48 @@ async def _process_pulled_events(
                 notification to clients, and validation of device keys.)
         """
 
+        logger.info(
+            "backfill events=%s",
+            [
+                "event_id=%s,depth=%d,body=%s,prevs=%s\n"
+                % (
+                    event.event_id,
+                    event.depth,
+                    event.content.get("body", event.type),
+                    event.prev_event_ids(),
+                )
+                for event in events
+            ],
+        )
```

Comment:
As a separate thing from this PR and more of a Python curiosity: why do I see both double and single quotes when printing a list of strings?

Here is an example log line from this log command:

```
2021-10-30 05:18:48,933 - synapse.handlers.federation_event - 755 - INFO - GET-4 - backfill events=["event_id=$OJ6FaiUsFz8OkdT6XkgXQUBRlJDXeyElHUilcLTA9UA,depth=11,body=Message 2,prevs=['$pMzSCdz615yH8cH9ptrTgrp0Eu3rI_YhAL4rA-fjfuc']\n", "event_id=$pMzSCdz615yH8cH9ptrTgrp0Eu3rI_YhAL4rA-fjfuc,depth=10,body=Message 1,prevs=['$_sn0C_7EIBCwtwQ7s0S9sPcQPmGw9nIWClzMfnqq2ss']\n", "event_id=$_sn0C_7EIBCwtwQ7s0S9sPcQPmGw9nIWClzMfnqq2ss,depth=9,body=Message 0,prevs=['$lqdI3q167_OCi4Il9_u2qftWSvh5hA3w2848QzrAs48']\n", "event_id=$fwHzQR3iUz7_WylSSbPnATXhw4vZzVL8IZDmgadzhI8,depth=9,body=org.matrix.msc2716.insertion,prevs=['$lqdI3q167_OCi4Il9_u2qftWSvh5hA3w2848QzrAs48']\n", "event_id=$j3NL0JeBa62l7zGjZkUsM2cpHvu-rX8_abtqfDSKQwE,depth=9,body=org.matrix.msc2716.batch,prevs=['$wruVNql_JRfLcg9FYNFLCWSpue_rSIinJy3N9sE0BtI']\n", "event_id=$wruVNql_JRfLcg9FYNFLCWSpue_rSIinJy3N9sE0BtI,depth=9,body=Historical 1 (batch=0),prevs=['$QiFiGh_BQJwoTE5tUfb6JIhiT3oeZ4BRSPAm5j0nMMk']\n", "event_id=$QiFiGh_BQJwoTE5tUfb6JIhiT3oeZ4BRSPAm5j0nMMk,depth=9,body=Historical 0 (batch=0),prevs=['$65Mj6CB-fBnvurwdZdjQVNLhM9o3XIqPlQOM-lp7jtE']\n", "event_id=$65Mj6CB-fBnvurwdZdjQVNLhM9o3XIqPlQOM-lp7jtE,depth=9,body=org.matrix.msc2716.insertion,prevs=['$hl13uJOaMDtfD26hxkdiHRusmO1f7uySwSaHKBQaBwI']\n", "event_id=$lqdI3q167_OCi4Il9_u2qftWSvh5hA3w2848QzrAs48,depth=8,body=Message 0,prevs=['$yI2LLl2uhM4bxpWcyxjv8XTXwHqSh1t4YZC3AoepBWk']\n", "event_id=$yI2LLl2uhM4bxpWcyxjv8XTXwHqSh1t4YZC3AoepBWk,depth=7,body=m.room.member,prevs=['$U8e-J6fb9vgqaJI8Y543IWSdWi7aeJmcbNXV7iCi9eo']\n", "event_id=$U8e-J6fb9vgqaJI8Y543IWSdWi7aeJmcbNXV7iCi9eo,depth=6,body=m.room.name,prevs=['$Mlr9i6GGyCSgkEkXeRKfsSrdIDk0tX0c9ALXzXUFOO4']\n", "event_id=$Mlr9i6GGyCSgkEkXeRKfsSrdIDk0tX0c9ALXzXUFOO4,depth=5,body=m.room.history_visibility,prevs=['$OYvnG2tvWS4SW0turXBdOP4l18hSwGI3uYUWkJRWTBg']\n", "event_id=$OYvnG2tvWS4SW0turXBdOP4l18hSwGI3uYUWkJRWTBg,depth=4,body=m.room.join_rules,prevs=['$D70kXVUeutU3UUNJbsqxP2VPjEzHxzNtC9RcIKwgaSY']\n", "event_id=$D70kXVUeutU3UUNJbsqxP2VPjEzHxzNtC9RcIKwgaSY,depth=3,body=m.room.power_levels,prevs=['$ceTe2BhJxa63HDcdgR-nS2H1xZkV7o9rIxPAIKE1IfQ']\n", "event_id=$ceTe2BhJxa63HDcdgR-nS2H1xZkV7o9rIxPAIKE1IfQ,depth=2,body=m.room.member,prevs=['$XTXwGqYulHWFXdmu7qQJdm2aTQjEFc15jyt-hB3_tuw']\n", 'event_id=$hl13uJOaMDtfD26hxkdiHRusmO1f7uySwSaHKBQaBwI,depth=1,body=m.room.member,prevs=[]\n', 'event_id=$XTXwGqYulHWFXdmu7qQJdm2aTQjEFc15jyt-hB3_tuw,depth=1,body=m.room.create,prevs=[]\n']
```

Making it more readable (notice the last two entries use single quotes):

```
backfill events=[
  "event_id=$OJ6FaiUsFz8OkdT6XkgXQUBRlJDXeyElHUilcLTA9UA,depth=11,body=Message 2,prevs=['$pMzSCdz615yH8cH9ptrTgrp0Eu3rI_YhAL4rA-fjfuc']",
  "event_id=$pMzSCdz615yH8cH9ptrTgrp0Eu3rI_YhAL4rA-fjfuc,depth=10,body=Message 1,prevs=['$_sn0C_7EIBCwtwQ7s0S9sPcQPmGw9nIWClzMfnqq2ss']",
  "event_id=$_sn0C_7EIBCwtwQ7s0S9sPcQPmGw9nIWClzMfnqq2ss,depth=9,body=Message 0,prevs=['$lqdI3q167_OCi4Il9_u2qftWSvh5hA3w2848QzrAs48']",
  "event_id=$fwHzQR3iUz7_WylSSbPnATXhw4vZzVL8IZDmgadzhI8,depth=9,body=org.matrix.msc2716.insertion,prevs=['$lqdI3q167_OCi4Il9_u2qftWSvh5hA3w2848QzrAs48']",
  "event_id=$j3NL0JeBa62l7zGjZkUsM2cpHvu-rX8_abtqfDSKQwE,depth=9,body=org.matrix.msc2716.batch,prevs=['$wruVNql_JRfLcg9FYNFLCWSpue_rSIinJy3N9sE0BtI']",
  "event_id=$wruVNql_JRfLcg9FYNFLCWSpue_rSIinJy3N9sE0BtI,depth=9,body=Historical 1 (batch=0),prevs=['$QiFiGh_BQJwoTE5tUfb6JIhiT3oeZ4BRSPAm5j0nMMk']",
  "event_id=$QiFiGh_BQJwoTE5tUfb6JIhiT3oeZ4BRSPAm5j0nMMk,depth=9,body=Historical 0 (batch=0),prevs=['$65Mj6CB-fBnvurwdZdjQVNLhM9o3XIqPlQOM-lp7jtE']",
  "event_id=$65Mj6CB-fBnvurwdZdjQVNLhM9o3XIqPlQOM-lp7jtE,depth=9,body=org.matrix.msc2716.insertion,prevs=['$hl13uJOaMDtfD26hxkdiHRusmO1f7uySwSaHKBQaBwI']",
  "event_id=$lqdI3q167_OCi4Il9_u2qftWSvh5hA3w2848QzrAs48,depth=8,body=Message 0,prevs=['$yI2LLl2uhM4bxpWcyxjv8XTXwHqSh1t4YZC3AoepBWk']",
  "event_id=$yI2LLl2uhM4bxpWcyxjv8XTXwHqSh1t4YZC3AoepBWk,depth=7,body=m.room.member,prevs=['$U8e-J6fb9vgqaJI8Y543IWSdWi7aeJmcbNXV7iCi9eo']",
  "event_id=$U8e-J6fb9vgqaJI8Y543IWSdWi7aeJmcbNXV7iCi9eo,depth=6,body=m.room.name,prevs=['$Mlr9i6GGyCSgkEkXeRKfsSrdIDk0tX0c9ALXzXUFOO4']",
  "event_id=$Mlr9i6GGyCSgkEkXeRKfsSrdIDk0tX0c9ALXzXUFOO4,depth=5,body=m.room.history_visibility,prevs=['$OYvnG2tvWS4SW0turXBdOP4l18hSwGI3uYUWkJRWTBg']",
  "event_id=$OYvnG2tvWS4SW0turXBdOP4l18hSwGI3uYUWkJRWTBg,depth=4,body=m.room.join_rules,prevs=['$D70kXVUeutU3UUNJbsqxP2VPjEzHxzNtC9RcIKwgaSY']",
  "event_id=$D70kXVUeutU3UUNJbsqxP2VPjEzHxzNtC9RcIKwgaSY,depth=3,body=m.room.power_levels,prevs=['$ceTe2BhJxa63HDcdgR-nS2H1xZkV7o9rIxPAIKE1IfQ']",
  "event_id=$ceTe2BhJxa63HDcdgR-nS2H1xZkV7o9rIxPAIKE1IfQ,depth=2,body=m.room.member,prevs=['$XTXwGqYulHWFXdmu7qQJdm2aTQjEFc15jyt-hB3_tuw']",
  'event_id=$hl13uJOaMDtfD26hxkdiHRusmO1f7uySwSaHKBQaBwI,depth=1,body=m.room.member,prevs=[]',
  'event_id=$XTXwGqYulHWFXdmu7qQJdm2aTQjEFc15jyt-hB3_tuw,depth=1,body=m.room.create,prevs=[]'
]
```

Reply:
IIRC, Python will default to displaying a string literal using single quotes. But if that string contains a `'` codepoints within it, it'll use `"` to delimit the string. If that string contains both `'` and `"` codepoints within it, it'll delimit with single quotes and escape any contained single quotes.

```python
>>> "a"
'a'
>>> "a's"
"a's"
>>> "a'\"s"
'a\'"s'
```

(well, I guess strictly speaking this might be a CPython implementation detail rather than a formally specced part of the language.)

Reply:
Thank you @DMRobertson 🙂 that lines up!